File: src/hetzner_code/state.py
"""Local state: one JSON file per instance under ~/.hetzner-code/instances/.

This is the only source of truth hcode trusts for "what did I create."
`status --reconcile` cross-checks it against hcloud/GitHub directly,
since a crash between steps can leave this file out of sync with the
real world — see hetzner_code.commands.status.
"""

from __future__ import annotations

import json
from dataclasses import asdict, dataclass, field

from hetzner_code.config import INSTANCES_DIR
from hetzner_code.run import HcodeError


@dataclass
class Repo:
    url: str  # the original git@github.com:owner/repo.git the user passed
    owner: str
    name: str  # short repo name, also the directory name on the box
    branch: str | None
    deploy_key_id: str  # GitHub's numeric id for this deploy key, for delete
    deploy_key_title: str


@dataclass
class Instance:
    name: str
    server_id: str
    ip: str
    type: str
    location: str
    login_key: str  # the hcloud SSH key *name* injected into the box
    login_key_path: (
        str  # local private key path used to actually SSH in as that identity
    )
    created_at: (
        str  # ISO 8601, set by the caller (Date.now() equivalents forbidden here)
    )
    repos: list[Repo] = field(default_factory=list)

    def to_json(self) -> str:
        return json.dumps(asdict(self), indent=2)

    @classmethod
    def from_json(cls, text: str) -> Instance:
        data = json.loads(text)
        data["repos"] = [Repo(**r) for r in data.get("repos", [])]
        return cls(**data)


def _instance_dir(name: str):
    return INSTANCES_DIR / name


def _meta_path(name: str):
    return _instance_dir(name) / "meta.json"


def save(instance: Instance) -> None:
    _instance_dir(instance.name).mkdir(parents=True, exist_ok=True)
    _meta_path(instance.name).write_text(instance.to_json())


def load(name: str) -> Instance:
    path = _meta_path(name)
    if not path.exists():
        raise HcodeError(f"no tracked instance named '{name}' (see `hcode status`)")
    return Instance.from_json(path.read_text())


def exists(name: str) -> bool:
    return _meta_path(name).exists()


def list_all() -> list[Instance]:
    if not INSTANCES_DIR.exists():
        return []
    out = []
    for child in sorted(INSTANCES_DIR.iterdir()):
        meta = child / "meta.json"
        if meta.exists():
            out.append(Instance.from_json(meta.read_text()))
    return out


def delete(name: str) -> None:
    import shutil

    d = _instance_dir(name)
    if d.exists():
        shutil.rmtree(d)


def key_dir(instance_name: str, repo_name: str):
    """Where a repo's local keypair lives before/while it's being
    uploaded. The private half is deleted after a successful scp —
    see hetzner_code.keys.forget_private_key."""
    return _instance_dir(instance_name) / "keys" / repo_name

File: internal/state/state.go
// Package state is local state: one JSON file per instance under
// ~/.hetzner-code/instances/.
//
// This is the only source of truth hcode trusts for "what did I
// create." `status --reconcile` cross-checks it against hcloud/GitHub
// directly, since a crash between steps can leave this file out of sync
// with the real world — see internal/commands' status command.
package state

import (
	"encoding/json"
	"os"
	"path/filepath"
	"sort"

	"github.com/dikee/hetzner-code/internal/config"
	"github.com/dikee/hetzner-code/internal/run"
)

// Repo is one codebase cloned onto an instance.
type Repo struct {
	URL            string   `json:"url"` // the original git@github.com:owner/repo.git the user passed
	Owner          string   `json:"owner"`
	Name           string   `json:"name"` // short repo name, also the directory name on the box
	Branch         *string  `json:"branch"`
	DeployKeyID    string   `json:"deploy_key_id"` // GitHub's numeric id for this deploy key, for delete
	DeployKeyTitle string   `json:"deploy_key_title"`
	Worktrees      []string `json:"worktrees"` // labels: "cc2", "cc3", ...
}

// Instance is one Hetzner box hcode created and is tracking.
type Instance struct {
	Name         string  `json:"name"`
	ServerID     string  `json:"server_id"`
	IP           string  `json:"ip"`
	Type         string  `json:"type"`
	Location     string  `json:"location"`
	LoginKey     string  `json:"login_key"`      // the hcloud SSH key *name* injected into the box
	LoginKeyPath string  `json:"login_key_path"` // local private key path used to actually SSH in as that identity
	CreatedAt    string  `json:"created_at"`     // ISO 8601, set by the caller
	Repos        []Repo  `json:"repos"`
	OpsDir       *string `json:"ops_dir"`       // remote path, if --ops-dir was copied up at create
	OpsDirLocal  *string `json:"ops_dir_local"` // its origin on your laptop — destroy syncs back here
}

func instanceDir(name string) string {
	return filepath.Join(config.InstancesDir, name)
}

func metaPath(name string) string {
	return filepath.Join(instanceDir(name), "meta.json")
}

// Save writes instance to its meta.json, creating the instance
// directory if needed.
func Save(instance Instance) error {
	for i := range instance.Repos {
		if instance.Repos[i].Worktrees == nil {
			instance.Repos[i].Worktrees = []string{}
		}
	}
	if instance.Repos == nil {
		instance.Repos = []Repo{}
	}
	if err := os.MkdirAll(instanceDir(instance.Name), 0o755); err != nil {
		return err
	}
	data, err := json.MarshalIndent(instance, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(metaPath(instance.Name), data, 0o644)
}

// Load reads a tracked instance by name.
func Load(name string) (Instance, error) {
	path := metaPath(name)
	data, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return Instance{}, run.Errorf("no tracked instance named '%s' (see `hcode status`)", name)
		}
		return Instance{}, err
	}
	var instance Instance
	if err := json.Unmarshal(data, &instance); err != nil {
		return Instance{}, err
	}
	return instance, nil
}

// Exists reports whether an instance by this name is already tracked.
func Exists(name string) bool {
	_, err := os.Stat(metaPath(name))
	return err == nil
}

// ListAll returns every tracked instance, sorted by name.
func ListAll() ([]Instance, error) {
	entries, err := os.ReadDir(config.InstancesDir)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}
	names := make([]string, 0, len(entries))
	for _, e := range entries {
		names = append(names, e.Name())
	}
	sort.Strings(names)

	var out []Instance
	for _, name := range names {
		if _, err := os.Stat(metaPath(name)); err == nil {
			instance, err := Load(name)
			if err != nil {
				return nil, err
			}
			out = append(out, instance)
		}
	}
	return out, nil
}

// Delete removes an instance's local state directory entirely.
func Delete(name string) error {
	d := instanceDir(name)
	if _, err := os.Stat(d); err == nil {
		return os.RemoveAll(d)
	}
	return nil
}

// KeyDir is where a repo's local keypair lives before/while it's being
// uploaded. The private half is deleted after a successful scp — see
// internal/keys.ForgetPrivateKey.
func KeyDir(instanceName, repoName string) string {
	return filepath.Join(instanceDir(instanceName), "keys", repoName)
}
